src/emulator/crypto/src/crypto_operations_factory.cpp

```diff
@@ -0,0 +1,17 @@
+#include <stdexcept>
+
+#include <crypto/crypto_operations_factory.h>
+#include <crypto/libtomcrypt_crypto_operations.h>
+
+#include <memory>
+
+std::shared_ptr<ICryptoOperations> CryptoOperationsFactory::create(CryptoOperationsTypes type)
```

Comment:
why not `std::unique_ptr`?

Reply:
Ask motoharu, I do not write this

Reply:
@motoharu-gosuto

Reply:
@VelocityRa 
Because crypto interface is passed everywhere. Copy constructor of unique_ptr is not accessable by design. You will either pass it as a reference (which is meaningless). Or only store it in host state and never pass it as argument.

Reply:
> (which is meaningless)

Can you elaborate? You pass it as a pointer everywhere, copying it everywhere for no reason (as far as I can tell). Why not pass as a referece and have only one defined place of ownership.

Reply:
@VelocityRa 
Because this is how you are supposed to use smart pointers?
They are designed to be an reference counted objects.
If you pass any smart pointer by reference - you are bypassing the core mechanism of these classes. Additionally if you are passing unique_ptr by reference - you are bypassing relationship control mechanism.

Reply:
>Because this is how you are supposed to use smart pointers?
I think you've misunderstood the goal of `shared_ptr` then.

In my opinion (and definitely not only mine) `shared_ptr` should be used when absolutely necessary.
If you say it's definitely neccesary here, fine by me, but passing smart pointers by reference is not as unheard of as you make it out to be.

Reply:
@VelocityRa 
I dont know man. As I said using references for smart pointers sounds completely weird to me.
And I am definitely not a fan of move semantics. Especially with things like unique_ptr.
If you want - you may try to adjust the code to use unique_ptr. I am not against it.

Reply:
You're probably right, I haven't thought this through very much. Maybe I'll try to mess around with it after it's merged, not very important.

Reply:
I'm sure it's fine, but I think it's strange to introduce a runtime cost (virtual functions) and the associated pointer management headaches when the implementation doesn't actually change at runtime.

I would probably have just declared the crypto functions and expect them to be provided.

Reply:
Depends on if we would need to switch to another crypto library or not. It gives you just enough flexibility. Not sure what you mean by :
'I would probably have just declared the crypto functions and expect them to be provided.'
Just declarations without a class?

Reply:
Yeah, and just let the linker sort it out.

Virtual functions are for dynamic dispatch -- implementations changing at runtime. We don't need that.

Reply:
I dont know. This sounds so much C. 
Also this was supposed to be integrated with pfs stuff in the future. 
Not sure how we will handle it without a crypto class.

Reply:
>This sounds so much C.

have you looked at the rest of the project? :p